src/C4EnergyBars.h

```diff
@@ -0,0 +1,263 @@
+/*
+ * LegacyClonk
+ *
+ * Copyright (c) RedWolf Design
+ * Copyright (c) 2017-2019, The LegacyClonk Team and contributors
+ *
+ * Distributed under the terms of the ISC license; see accompanying file
+ * "COPYING" for details.
+ *
+ * "Clonk" is a registered trademark of Matthes Bender, used with permission.
+ * See accompanying file "TRADEMARK" for details.
+ *
+ * To redistribute this file separately, substitute the full license texts
+ * for the above references.
+ */
+
+/* Allows to define custom energy bars */
+
+#pragma once
+
+class C4EnergyBars;
+
+#include "C4FacetEx.h"
+#include "C4Id.h"
+#include "C4Object.h"
+
+#include <unordered_map>
+#include <vector>
+#include <memory>
+#include <set>
+#include <string>
+
+class C4EnergyBar
+{
+public:
+	int32_t value;
+	int32_t max;
+	bool visible;
+
+	C4EnergyBar(): value(0), max(1000000), visible(true) {}
+	C4EnergyBar(int32_t _value, int32_t _max, bool _visible);
+	bool operator==(const C4EnergyBar &rhs) const {
+		return value == rhs.value && max == rhs.max && visible == rhs.visible;
+	}
+
+	void CompileFunc(StdCompiler *pComp) {
+		pComp->Value(mkNamingAdapt(value, "Value", 0));
+		pComp->Value(mkNamingAdapt(max, "Max", 1000000));
+		pComp->Value(mkNamingAdapt(visible, "Visible", true));
+	}
+};
+
+class C4EnergyBarsDef;
+class C4Object;
+
+class C4EnergyBars
+{
+public:
+	std::shared_ptr<C4EnergyBarsDef> def;
+	std::vector<C4EnergyBar> values;
+
+	C4EnergyBars(std::shared_ptr<C4EnergyBarsDef> _def);
+
+	void DrawEnergyBars(C4Facet &cgo, C4Object &obj);
+	void SetEnergyBar(std::string name, int32_t value, int32_t max = 0);
+	void SetEnergyBarVisible(std::string name, bool visible);
+
+private:
+	C4EnergyBar* BarVal(std::string name);
+};
+
+class C4EnergyBarDef
+{
+public:
+	enum Physical : int32_t {
+		EBP_None   = 0,
+		EBP_Energy = 1,
+		EBP_Magic  = 2,
+		EBP_Breath = 3,
+		EBP_All = EBP_Energy | EBP_Magic | EBP_Breath
+	};
+	enum Hide : int32_t {
+		EBH_Never = 0,
+		EBH_Empty = 1,
+		EBH_Full = 2,
+		EBH_EmptyFull = EBH_Empty | EBH_Full,
+		EBH_HideHUDBars = 512,
+		EBH_All = EBH_EmptyFull | EBH_HideHUDBars
+	};
```

Comment:
I did it like that because I saw it being done that way in C4Def, from that code it seemed like the codebase prefers int32_t. (but that could simply be cause the code is old...)
At first I tried using enum types directly but the code was way more awkward in comparison to the relative simplicity of doing it the C4Def way `C4Def::HideBar` `C4Def::HideHud`. Especially CompileFunc seemed more complicated with enum only.

That said, could be that there are ways to use enums only, without having to awkwardly define lots of type conversion / CompileFunc code.

Reply:
I already forgot that I did it that way for the Hide stuff. Now I am wondering why.
Maybe @Fulgen301 has an opinion on that?

Reply:
The conclusion of a short discussion is, that we don’t want to specify the underlying type for enums that are not serialized in binary (which is neither the case here, nor for the Hide stuff). Instead let the compiler choose an appropriate type.